packages/protocol/contracts/L1/libs/LibL2Tokenomics.sol

```diff
@@ -0,0 +1,128 @@
+// SPDX-License-Identifier: MIT
+//  _____     _ _         _         _
+// |_   _|_ _(_) |_____  | |   __ _| |__ ___
+//   | |/ _` | | / / _ \ | |__/ _` | '_ (_-<
+//   |_|\__,_|_|_\_\___/ |____\__,_|_.__/__/
+
+pragma solidity ^0.8.18;
+
+import {AddressResolver} from "../../common/AddressResolver.sol";
+import {LibMath} from "../../libs/LibMath.sol";
+import {LibFixedPointMath} from "../../thirdparty/LibFixedPointMath.sol";
+import {
+    SafeCastUpgradeable
+} from "@openzeppelin/contracts-upgradeable/utils/math/SafeCastUpgradeable.sol";
+import {TaikoData} from "../TaikoData.sol";
+
+library LibL2Tokenomics {
+    using LibMath for uint256;
+    using SafeCastUpgradeable for uint256;
+
+    error L1_1559_GAS_CHANGE_MISMATCH(uint64 expectedRatio, uint64 actualRatio);
+    error L1_1559_X_SCALE_TOO_LARGE();
+    error L1_1559_Y_SCALE_TOO_LARGE();
+    error L1_OUT_OF_BLOCK_SPACE();
+
+    function getL2Basefee(
+        TaikoData.State storage state,
+        TaikoData.Config memory config,
+        uint32 gasLimit
+    ) internal view returns (uint64 basefee, uint64 newGasExcess) {
+        if (config.gasIssuedPerSecond == 0) {
+            // L2 1559 disabled
+            return (0, 0);
+        }
+
+        unchecked {
+            uint256 reduced = (block.timestamp - state.lastProposedAt) *
+                config.gasIssuedPerSecond;
+            newGasExcess = uint64(reduced.max(state.l2GasExcess) - reduced);
+        }
+
+        uint256 _basefee = calcL2Basefee({
+            l2GasExcess: newGasExcess,
+            xscale: state.l2Xscale,
+            yscale: uint256(state.l2Yscale) << 64,
+            gasAmount: gasLimit
```

Comment:
So this implies that a proposer needs to select a gasLimit of a block beforehand, then choose the transactions in mempool to try to fill that gas limit right? And our curcuits need to "filter" those transactions in a block that do not have a high enough `tx.maxFeePerGas`?